python/cudf/cudf/core/column/numerical.py

```diff
@@ -510,13 +510,52 @@ def can_cast_safely(self, to_dtype):
                     info = np.finfo(to_dtype)
                 elif "int" in to_dtype.name:
                     info = np.iinfo(to_dtype)
-                min_, max_ = info.min, info.max
+                lower_, upper_ = info.min, info.max
+
+                if self.dtype.kind == "f":
+                    # Exclude 'np.inf', '-np.inf'
+                    s = cudf.Series(self)
+                    non_infs = s[
+                        ((s == np.inf) | (s == -np.inf)).logical_not()
+                    ]
+                    col = non_infs._column
+                else:
+                    col = self
+
+                min_ = col.min()
+                if min_ is np.nan:
+                    # Column contains only infs
+                    return True
 
-                if (self.min() > min_) and (self.max() < max_):
+                max_ = col.max()
+                if (min_ >= lower_) and (max_ < upper_):
                     return True
                 else:
                     return False
 
+        # want to cast uint to int
+        elif self.dtype.kind == "i" and to_dtype.kind == "u":
+            i_max_ = np.iinfo(self.dtype).max
+            u_max_ = np.iinfo(to_dtype).max
+
+            if self.min() >= 0:
+                if i_max_ <= u_max_:
+                    return True
+                if self.max() < u_max_:
+                    return True
```

Comment:
Do we want to do it based on only the type, or should we introspect the data to see if the max / min of the data fit into the data type?

Reply:
This does introspect the data, the previous line is a shortcut - if the integer type max is smaller, no need to check the type; otherwise the introspect will be performed.